### ISSUE TITLE
BleError: Device [ID] was disconnected on Android

### ISSUE THREAD

**Issue by bkbooth** [LedgerHQ/ledgerjs#511]
I'm trying to use `react-native-hw-transport-ble` and on Android it's throwing lots of `BleError: Device [ID] was disconnected` errors. Unfortunately the errors are being thrown from somewhere that I don't seem to be able to catch. On iOS everything is working fine.

I can replicate this by finding a device with `TransportBLE.listen()` and opening it with `TransportBLE.open()`. Then if I go back to the previous screen, then forward to the BLE connection screen again, I'll get one of the `BleError`s thrown (from somewhere) when I try to connect again, but my application code will actually return with a valid transport from `TransportBLE.open()` despite the error being thrown.

I've tried explicitly disconnecting with `TransportBLE.disconnect()` before attempting `TransportBLE.open()` (and catching errors if the device isn't already connected), but the same thing happens.

I've noticed that if I inspect `transport.device._manager`, the `_activeSubscriptions` list grows each time I try to reconnect, so I'm suspecting the error is being thrown from one of those subscriptions, but I'm not explicitly subscribing to the `BleManager` in my application code.

There's a chance I'm doing something dumb/wrong here but it's all running well on iOS and I'm running out of ideas with this. Any ideas or suggestions are much appreciated.

**Comment by gre** [comment]
it sounds like you are stuck in a "bluetooth bonding issue" that exists in current Nano X firmware and requires you to Reset the BLE connection => https://support.ledger.com/hc/en-us/articles/360025864773-Fix-Bluetooth-issues

what is the device state when you are initially pairing the device? I'm aware of issues if you do it from within an app, **the first bonding need to be done on the dashboard of the device**. in Ledger Live Mobile we "remember" on our app that connected device so we can mitigate the problem by having a "Pair a new device" flow that will guide user to connect a new device when device is on "dashboard". Then once you know the device id (`event.descriptor` of listen), you can just open(id) it

**Comment by bkbooth** [comment]
Thanks for your reply @gre. I've been playing around with this some more and I still can't work it out.

To be clear, I haven't attempted doing the initial device bonding from the dashboard. I'll play around with that next but it fairly majorly affects our workflow if we need to guide users through pairing on the dashboard, then open the app, then connect again. Our current workflow is smooth and easy for users on iOS, it'd be a shame to lose some of that because of bluetooth issues which seem to be Android-only.

Our ideal workflow is that the user has the device unlocked with the right app already open when we're trying to connect, then with a single press of a button from the user we can call `.open()` to create the transport, and start reading addresses from the app. But ideally we'd like it to not matter if the app is open or not for initial bonding, we show a message asking them to open the app on the device if reading addresses fails.

Just to be clear, you're saying that we should:

1. For first pairing/connection, use `TransportBLE.listen()` to look for nearby devices with event type `add` (the device should be on the dashboard).
2. Choose one of the `add`ed devices.
3. Call `TransportBLE.open(event.descriptor.id)` to create the initial bond.
4. Call `TransportBLE.disconnect(event.descriptor.id)` to save/remember the bonding.
5. Store and return the `event.descriptor.id` for subsequent pairing requests to that device.

We're currently not doing steps 3 & 4 from that, we return the device ID and start trying to use it immediately.

**Comment by bkbooth** [comment]
Here's a snippet for where I'm attempting the initial device pairing/connection (I keep resetting pairings between testing to be sure). This is the same whether I'm on the device dashboard, or inside an app on the device.

For this code:
```
try {
  // event is a TransportBLE.listen() event filtered to type 'add'
  const device = event.descriptor;
  console.log('found ble device', device.id);
  // Attempt to connect then disconnect to pair initial bond
  console.log('connecting...');
  await TransportBLE.open(device.id);
  console.log('disconnecting...');
  await TransportBLE.disconnect(device.id);
  console.log('returning...');
  return device;
} catch (error) {
  console.log('something went wrong', { error });
  return null;
}
```

I'm seeing this in the log:

```
found ble device 12:34:56:78:90:00
connecting...
BleError: Device 12:34:56:78:90:00 was disconnected
disconnecting...
returning...
```

Then after returning the device ID and attempting to connect to use it in earnest (`Transport.open(device.id)`), I'll get 2 of the `BleError`s. I can't seem to catch these `BleError`s, they're coming from outside of my application code. In development mode they show the red full screen error messages, but in release builds they cause the whole app to crash.

**Comment by gre** [comment]
Strange.
Are you able to make this device work with Ledger Live on the same device you are trying?

Ledger live mobile implementation is basically this: https://github.com/LedgerHQ/ledger-live-mobile/blob/develop/src/screens/PairDevices/index.js#L94

We do a bit more things between open and close, maybe you can try to fetch the device info (which is a good sanity check to make sure device is effectively on dashboard)

https://github.com/LedgerHQ/ledger-live-common/blob/master/src/hw/getVersion.js

**Comment by bkbooth** [comment]
Yep, I can add the device with Ledger Live and it is (seems to be?) working correctly.

I just noticed that I'm using `react-native-ble-plx@2.0.0`, but Ledger Live is using v1.1.1 (and that `@ledgerhq/react-native-hw-transport-ble` has a peerDependency of `react-native-ble-plx@^1.0.1`). I'm going to try rolling back my `react-native-ble-plx` version 🤞

**Comment by bkbooth** [comment]
Downgrading `react-native-ble-plx` didn't work because I'm on a newer version of React Native than v1.1.1 supports, but they just released [v2.0.1](https://github.com/Polidea/react-native-ble-plx/releases/tag/2.0.1) yesterday which appears to fix the issue on Android 🎉

Thanks for your help @gre! It's very much appreciated, even though the problem ended up being a bug in an upstream dependency. Looking through the Ledger Live code was a point in the right direction.

For what it's worth, making the initial pairing/connection appears to be working fine even if the Nano X has an app open.
